subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -0,0 +1,147 @@
+        RecompilationSpec spec = new RecompilationSpec();
+        if (sourceFileClassNameConverter.isEmpty()) {
+            spec.setFullRebuildCause("no source class mapping file found", null);
+            return spec;
+        }
+
+        processClasspathChanges(current, previous, spec);
+        processOtherChanges(previous, spec);
+
+        spec.getClassesToProcess().addAll(previous.getTypesToReprocess());
+        return spec;
+    }
+
+    @Override
+    public void initializeCompilation(JavaCompileSpec spec, RecompilationSpec recompilationSpec) {
+        if (!recompilationSpec.isBuildNeeded()) {
+            spec.setSourceFiles(Collections.emptySet());
+            spec.setClasses(Collections.emptySet());
+            return;
+        }
+
+        Factory<PatternSet> patternSetFactory = fileOperations.getFileResolver().getPatternSetFactory();
+        PatternSet classesToDelete = patternSetFactory.create();
+        PatternSet filesToRecompile = patternSetFactory.create();
+
+        prepareFilePatterns(recompilationSpec.getFilesToCompile(), classesToDelete, filesToRecompile);
+
+        spec.setSourceFiles(sourceTree.matching(filesToRecompile));
+        includePreviousCompilationOutputOnClasspath(spec);
+        addClassesToProcess(spec, recompilationSpec);
+
+        deleteStaleFilesIn(classesToDelete, spec.getDestinationDir());
+    }
+
+    private void prepareFilePatterns(Set<File> filesToCompile, PatternSet classesToDelete, PatternSet filesToRecompilePatterns) {
+        for (File file : filesToCompile) {
+            filesToRecompilePatterns.include(relativize(file));
+
+            Collection<String> classes = sourceFileClassNameConverter.getClassNames(file);
+            for (String staleClass : classes) {
+                String path = staleClass.replaceAll("\\.", "/");
+                classesToDelete.include(path.concat(".class"));
+            }
+        }
+    }
+
+    private String relativize(File sourceFile) {
+        List<File> dirs = sourceDirs.getSourceRoots();
+        for (File sourceDir : dirs) {
+            if (sourceFile.getAbsolutePath().startsWith(sourceDir.getAbsolutePath())) {
+                return RelativePathUtil.relativePath(sourceDir, sourceFile);
+            }
+        }
+        throw new IllegalStateException("Not found " + sourceFile + " in source dirs: " + dirs);
+    }
+
+    private void processOtherChanges(PreviousCompilation previous, RecompilationSpec spec) {
+        if (spec.getFullRebuildCause() != null) {
+            return;
+        }
+        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous, sourceFileClassNameConverter);
+        for (FileChange fileChange : sourceChanges) {
+            if (spec.getFullRebuildCause() != null) {
+                return;
+            }
+            spec.getFilesToCompile().add(fileChange.getFile());
```

Comment:
I wonder if we need to ignore directories here. `InputChanges` will also report directories...

Note that most incremental compilation tests don't use any package names, so we don't have much test coverage in that area.